packages/@aws-cdk/aws-batch/lib/job-definition.ts

```diff
@@ -46,6 +53,28 @@ export enum LogDriver {
   SYSLOG = 'syslog'
 }
 
+/**
+ * An object representing the secret to expose to your container
+ */
+export interface SecretProperty {
+  /**
+   * The name of the secret.
+   */
+  readonly name: string;
+
+  /**
+   * A secret from secrets manager
+   * @default - one of secretsManagerValue or parametersStoreValue is required
+   */
+  readonly secretsManagerValue?: secretsmanager.ISecret;
+
+  /**
+   * A parameter from parameters store
+   * @default - one of secretsManagerValue or parametersStoreValue is required
+   */
+  readonly parametersStoreValue?: ssm.IParameter;
```

Comment:
To avoid this mutual exclusion, we normally follow a *union like* class pattern. 

```ts
readonly secret: ExposedSecret;
```

The caller would then do: `batch.ExposedSecret.fromSecretsManager()` or `batch.ExposedSecret.fromParametersStore`.  This way you won't have to do any runtime validation that only one these was passed. And since the property is now required, you also don't have to validate some property was given.

You can see an example here:

https://github.com/aws/aws-cdk/blob/60875a5d6cd93b7d6f0f9b1b94153f6d1f4766b1/packages/%40aws-cdk/aws-codebuild/lib/build-spec.ts#L6-L18

Reply:
@iliapolo How does this look?

```ts
import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
import * as ssm from '@aws-cdk/aws-ssm';

export interface SecretsManagerSecretProps {
  name: string;
  value: secretsmanager.ISecret;
}

export interface ParametersStoreSecretProps {
  name: string;
  value: ssm.IParameter;
}

export abstract class ExposedSecret {
  public static fromSecretsManager(props: SecretsManagerSecretProps): ExposedSecret {
    return new SecretsManagerSecret(props);
  }

  public static fromParametersStore(props: ParametersStoreSecretProps): ExposedSecret {
    return new ParametersStoreSecret(props);
  }

  public abstract secretName: string;
  public abstract secretArn: string;
}


class SecretsManagerSecret extends ExposedSecret {
  public secretName: string;
  public secretArn: string;

  constructor(props: SecretsManagerSecretProps) {
    super();
    this.secretName = props.name;
    this.secretArn = props.value.secretArn;
  }
}
class ParametersStoreSecret extends ExposedSecret {
  public secretName: string;
  public secretArn: string;

  constructor(props: ParametersStoreSecretProps) {
    super();
    this.secretName = props.name;
    this.secretArn = props.value.parameterArn;
  }
}
```